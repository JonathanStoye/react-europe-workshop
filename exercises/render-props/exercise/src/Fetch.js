import React, { Component } from "react";

class Fetch extends Component {
  state = {
    loading: true,
    error: null,
    data: null
  };

  componentDidMount() {
    fetch(this.props.url)
      .then(res => res.json())
      .catch(error => {
        this.setState({
          error,
          loading: false
        });
      });
  }

  render() {
    return this.props.children(this.state);
  }
}

export default Fetch;
