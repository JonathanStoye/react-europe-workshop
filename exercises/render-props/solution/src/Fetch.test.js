import React from "react";
import renderer from "react-test-renderer";
import Fetch from "./Fetch";

global.fetch = jest.fn();

describe("Fetch Commponent", () => {
  it("renders a snapshot that is good", () => {
    global.fetch.mockReturnValue(Promise.resolve({}));
    const tree = renderer
      .create(<Fetch url="https://something.com">{args => null}</Fetch>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("start with loading === true", () => {
    let result;
    const tree = renderer.create(
      <Fetch>
        {({ loading }) => {
          result = loading;
          return null;
        }}
      </Fetch>
    );
    expect(result).toBe(true);
  });

  it("should do data", done => {
    global.fetch.mockReturnValue(
      new Promise((resolve, reject) => {
        setTimeout(() => resolve({ json: () => "DATA" }), 1);
      })
    );
    let result;
    const tree = renderer.create(
      <Fetch>
        {({ data }) => {
          result = data;
          return null;
        }}
      </Fetch>
    );
    setTimeout(() => {
      expect(result).toBe("DATA");
      done();
    }, 2);
  });

  it("should do error", done => {
    global.fetch.mockReturnValue(
      new Promise((resolve, reject) => {
        setTimeout(() => reject("ERROR"), 1);
      })
    );
    let result;
    const tree = renderer.create(
      <Fetch>
        {({ error }) => {
          result = error;
          return null;
        }}
      </Fetch>
    );
    setTimeout(() => {
      expect(result).toBe("ERROR");
      done();
    }, 2);
  });
});
