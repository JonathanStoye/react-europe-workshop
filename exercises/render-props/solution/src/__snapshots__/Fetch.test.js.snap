// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch Commponent renders a snapshot that is good 1`] = `null`;
